src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -59,93 +61,113 @@ static void callback(ConstFSEventStreamRef streamRef,
+    invalidStateDetected = false;
+    CFMutableArrayRef rootsToWatch = CFArrayCreateMutable(NULL, 0, NULL);
     if (rootsToWatch == NULL) {
-        // nothing to watch, just return
-        return;
+        mark_failed_with_errno(env, "Could not allocate array to store roots to watch.", result);
+        return NULL;
+    }
+    int count = env->GetArrayLength(paths);
+    if (count == 0) {
+        mark_failed_with_errno(env, "No paths given to watch.", result);
+        return NULL;
+    }
+    for (int i = 0; i < count; i++) {
+        jstring path = (jstring) env->GetObjectArrayElement(paths, i);
+        char* pathString = java_to_char(env, path, result);
+        if (pathString == NULL) {
+            mark_failed_with_errno(env, "Could not allocate string to store root to watch.", result);
+            return NULL;
+        }
+        CFStringRef stringPath = CFStringCreateWithCString(NULL, pathString, kCFStringEncodingUTF8);
+        free(pathString);
+        if (stringPath == NULL) {
+            mark_failed_with_errno(env, "Could not create CFStringRef.", result);
+            return NULL;
+        }
+        CFArrayAppendValue(rootsToWatch, stringPath);
     }
-    CFAbsoluteTime latency = 0.3;  // Latency in seconds
 
-    watcherCallback = env->NewGlobalRef(javaCallback);
+    jobject watcherCallback = env->NewGlobalRef(javaCallback);
     if (watcherCallback == NULL) {
         mark_failed_with_errno(env, "Could not create global reference for callback.", result);
-        return;
+        return NULL;
     }
 
-    watcherStream = FSEventStreamCreate (
+    FSEventStreamContext context = {0, (void*) watcherCallback, NULL, NULL, NULL};
+    FSEventStreamRef watcherStream = FSEventStreamCreate (
                 NULL,
                 &callback,
-                NULL,
+                &context,
                 rootsToWatch,
                 kFSEventStreamEventIdSinceNow,
                 latency,
                 kFSEventStreamCreateFlagNoDefer);
     if (watcherStream == NULL) {
          mark_failed_with_errno(env, "Could not create FSEventStreamCreate to track changes.", result);
-         return;
+         return NULL;
     }
 
-    if (pthread_create(&watcherThread, NULL, EventProcessingThread, watcherStream) != 0) {
+    watch_details_t* details = (watch_details_t*)malloc(sizeof(watch_details_t));
+    details->rootsToWatch = rootsToWatch;
+    details->watcherStream = watcherStream;
+    details->watcherCallback = watcherCallback;
+
+    if (pthread_create(&(details->watcherThread), NULL, EventProcessingThread, details) != 0) {
         mark_failed_with_errno(env, "Could not create file watcher thread.", result);
-        return;
+        return NULL;
     }
 
+    // TODO Should this be somewhere global?
     int jvmStatus = env->GetJavaVM(&jvm);
     if (jvmStatus < 0) {
         mark_failed_with_errno(env, "Could not store jvm instance.", result);
-        return;
+        return NULL;
```

Comment:
should probably kill the watcher thread and free `details` here

Reply:
Fixed.